fendermint/vm/interpreter/src/chain.rs

```diff
@@ -58,22 +61,27 @@ pub type ChainMessageCheckRes = Result<SignedMessageCheckRes, IllegalMessage>;
 /// Interpreter working on chain messages; in the future it will schedule
 /// CID lookups to turn references into self-contained user or cross messages.
 #[derive(Clone)]
-pub struct ChainMessageInterpreter<I> {
+pub struct ChainMessageInterpreter<I, P> {
     inner: I,
+    parent_finality_provider: PhantomData<P>,
 }
 
-impl<I> ChainMessageInterpreter<I> {
+impl<I, P> ChainMessageInterpreter<I, P> {
     pub fn new(inner: I) -> Self {
-        Self { inner }
+        Self {
+            inner,
+            parent_finality_provider: Default::default(),
+        }
     }
 }
 
 #[async_trait]
-impl<I> ProposalInterpreter for ChainMessageInterpreter<I>
+impl<I, P> ProposalInterpreter for ChainMessageInterpreter<I, P>
 where
     I: Sync + Send,
+    P: ParentFinalityProvider + Send + Sync,
 {
-    type State = (CheckpointPool, Arc<InMemoryFinalityProvider>);
+    type State = (CheckpointPool, Arc<P>);
```

Comment:
Hm, this is getting a bit infectious. I reckon a `Toggle<InMemoryFinalityProvider>` we could avoid the extra `P` and the phantomdata and save a lot of extra `where P: ParentFinalityProvider + Send + Sync` along the way, and then it would only affect the traits where its' part of the state.